src/lib.rs

```diff
@@ -198,67 +287,200 @@ impl IRCompilerFactory {
 
-        let v = unsafe {
+        let v = NonNull::new(unsafe {
             self.funcs.IRCompilerAllocCompileAndLink(
-                self.me,
+                self.me.as_ptr(),
                 entry_point.as_ptr(),
-                input.me,
+                input.me.as_ptr(),
                 &mut error,
             )
-        };
+        });
 
         if error.is_null() {
             Ok(IRObject {
-                me: v,
+                me: v.unwrap(),
                 funcs: input.funcs.clone(),
             })
         } else {
-            panic!("{:?}", error);
+            let error = IRError::from_ptr(error, self.funcs.clone());
+            Err(CompilerError(error.code()))
+        }
+    }
+}
+
+struct IRError {
+    me: NonNull<bindings::IRError>,
+    funcs: Arc<bindings::metal_irconverter>,
+}
+
+impl Drop for IRError {
+    #[doc(alias = "IRErrorDestroy")]
+    fn drop(&mut self) {
+        unsafe { self.funcs.IRErrorDestroy(self.me.as_ptr()) };
+    }
+}
+
+impl IRError {
+    pub fn from_ptr(ptr: *mut bindings::IRError, funcs: Arc<bindings::metal_irconverter>) -> Self {
+        Self {
+            me: NonNull::new(ptr).unwrap(),
+            funcs,
+        }
+    }
+
+    #[doc(alias = "IRErrorGetCode")]
+    pub fn code(&self) -> IRErrorCode {
+        let code = unsafe { self.funcs.IRErrorGetCode(self.me.as_ptr()) };
+        match code {
+            0 => IRErrorCode::NoError,
+            1 => IRErrorCode::ShaderRequiresRootSignature,
+            2 => IRErrorCode::UnrecognizedRootSignatureDescriptor,
+            3 => IRErrorCode::UnrecognizedParameterTypeInRootSignature,
+            4 => IRErrorCode::ResourceNotReferencedByRootSignature,
+            5 => IRErrorCode::ShaderIncompatibleWithDualSourceBlending,
+            6 => IRErrorCode::UnsupportedWaveSize,
+            7 => IRErrorCode::UnsupportedInstruction,
+            8 => IRErrorCode::CompilationError,
+            9 => IRErrorCode::FailedToSynthesizeStageInFunction,
+            10 => IRErrorCode::FailedToSynthesizeStreamOutFunction,
+            11 => IRErrorCode::FailedToSynthesizeIndirectIntersectionFunction,
+            12 => IRErrorCode::UnableToVerifyModule,
+            13 => IRErrorCode::UnableToLinkModule,
+            14 => IRErrorCode::UnrecognizedDXILHeader,
+            15 => IRErrorCode::InvalidRaytracingAttribute,
+            16 => IRErrorCode::Unknown,
+            _ => panic!("Invalid error code"),
+        }
+    }
+
+    #[doc(alias = "IRErrorGetPayload")]
+    pub fn _payload(&self) -> CString {
+        unsafe {
+            // The documentation is inconsistent about this function.
+            // The docs say `You must cast this pointer to the appropriate error payload struct for the error code``
+            // but the example code just treats this as a char* like this: `printf("%s\n", (const char*)IRErrorGetPayload(pRootSigError));`
+            // Let's assume the example code is correct
```

Comment:
We should request a clarification that explains for each `IRErrorGetCode()` how the payload should be interpreted. Maybe only linking gives an accompanying error message.

Besides, you're never using nor exposing this payload to the user, why?